### PR TITLE
union data creates empty tables if none exist

### PR DESCRIPTION
**Pull request by fivetran-jamie**
**What change does this PR introduce?** 
 
- adds argument to union_data to handle when there's no table in any of the provided schemas/databases (by creating an empty table) 

https://github.com/fivetran/dbt_fivetran_utils/issues/99 (came out of https://github.com/fivetran/dbt_shopify_source/issues/57)

**If this PR introduces a new macro, how did you test the new macro?** 

on shopify

**If this PR introduces a modification to an existing macro, which packages is the macro currently present in and what steps were taken to test compatibility across packages?** 

Packages that use the `union_data` macro (found by searching `union_data` on github within the Fivetran org), and the impact of this change on them:
- Facebook pages
  - If a table doesn’t exist, in previous versions the package would have failed. Now it creates an empty ones and works with those
- Twitter organic  
  - If a table doesn’t exist, in previous versions the package would have failed. Now it creates an empty ones and works with those
- Instagram business
  - If a table doesn’t exist, in previous versions the package would have failed. Now it creates an empty ones and works with those
- Linkedin pages 
  - If a table doesn’t exist, in previous versions the package would have failed. Now it creates an empty ones and works with those
- Stripe
  - Instead of dynamically creating empty tables, Stripe currently has using__tablename variables to disable these. So for customers that have these set to false, nothing changes.
  - For customers that have these set to true, and the table exists, nothing changes. 
  - For customers that have these set to true, and the table doesn’t exist, we are resolving an error.
  - For tables that aren’t covered by the variables, if they don’t exist, they caused errors in previous versions. Now they won’t
- Xero
  - Instead of dynamically creating empty tables, xero currently has using__tablename variables to disable these (2). So for customers that have these set to false, nothing changes.
  - For customers that have these set to true, and the table exists, nothing changes. 
  - For customers that have these set to true, and the table doesn’t exist, we are resolving an error.
  - For tables that aren’t covered by the variables, if they don’t exist, they caused errors in previous versions. Now they won’t
- Klaviyo 
  - If a table doesn’t exist, in previous versions the package would have failed. Now it creates an empty ones and works with those

Packages upstream of the above:
- Shopify holistic reporting
- social media reporting

**Did you update the README to reflect the macro addition/modifications?** 

- [x] Yes
- [ ] No (provide further explanation)

### CODE REVIEW

**Comment by fivetran-jamie** [comment]
in a way the column_macro stuff isn't the 100% necessary within the context of our source packages, as the non-tmp staging model will call `fill_staging_column` itself and do the null-casting....

File: macros/union_data.sql
Comment (fivetran-joemarkiewicz) [inline macros/union_data.sql]: As I stated in the [shopify_source review](https://github.com/fivetran/dbt_shopify_source/pull/59), I think this is possibly duplicative. I think we can remove the column_macro since we can gain comfort that the staging model will ensure success and parity of fields.

Additionally, this results in less maintenance across the packages!

File: macros/union_data.sql
Comment (fivetran-joemarkiewicz) [inline macros/union_data.sql]: Same comment as above.

**Review by fivetran-joemarkiewicz — COMMENTED** [review]
@fivetran-jamie thanks for working through this unique issue and finding a solution that can be easily scaled out to all our packages! I did have one comment following my review of the shopify_source package and that is to consider removing the `column_macro` since this is not entirely necessary.

Let me know once that has been updated and I can do a final review.

File: macros/union_data.sql
Comment (fivetran-jamie) [reply macros/union_data.sql]: removed

File: macros/union_data.sql
Comment (fivetran-jamie) [reply macros/union_data.sql]: removed

**Comment by fivetran-sheringuyen** [comment]
Overall, this is a really neat solution @fivetran-jamie! 

As a reviewer, I personally felt very easily overwhelmed/lost with this macro over all so thank you for taking the time to review it with me. (But it makes me consider if we should create a separate issue to provide like working examples of these more complex macros so that it doesn't require the review/explanation again 🤔 ). 

I'm wondering if in addition to making sure the test doesn't fail, that we should output some kind of messaging to ensure that the user knows that their listed tables were not actually found and have yielded an empty table(s). I'm just thinking that previously, they would have known due to a failure in the macro and could have had an option to fix it on their side. 

And a side note, once this branch merges with `feature/unit-testing-phase1` buildkite should pass!

**Review by fivetran-sheringuyen — APPROVED** [review]

**Comment by fivetran-jamie** [comment]
i like that idea @fivetran-sheringuyen ! do you know if it's possible to add a custom message to a test output? 


also should i change the base branch of this PR to `feature/unit-testing-phase1`? or merge `feature/unit-testing-phase1` into my working branch? or wait for `feature/unit-testing-phase1` to merge into the release branch?  or something else altogether....

File: macros/union_data.sql
Comment (fivetran-joemarkiewicz) [inline macros/union_data.sql]: Now that I think about this, is "Warning" the right word here? A warning to me indicates something needs to be looked into, this may not 🤷. Is there a less severe word we could possibly use here?

Comment (fivetran-joemarkiewicz) [reply macros/union_data.sql]: Additionally, what are your thoughts in providing the variable to disable in the message itself? Just so customers are directly aware of it and can turn it off without needing to inspect our docs?

**Review by fivetran-joemarkiewicz — COMMENTED** [review]
@fivetran-jamie before approving this I still want to test it on the other packages with the union data functionality, but I did have a few comments below which I would like your input.

Additionally, will we need to add the README and DECISIONLOG entry you have in shopify_source to the other packages as well? My understanding is that we will, but wanted to confirm on your end before approving. This way we can make a plan for what needs to be updated.

File: macros/union_data.sql
Comment (fivetran-jamie) [reply macros/union_data.sql]: >Now that I think about this, is "Warning" the right word here? A warning to me indicates something needs to be looked into, this may not 🤷. Is there a less severe word we could possibly use here?

Yeah honestly i thought about "FYI" or "Heads up" lol

>Additionally, what are your thoughts in providing the variable to disable in the message itself? Just so customers are directly aware of it and can turn it off without needing to inspect our docs?

 I really like that idea!!

**Comment by fivetran-jamie** [comment]
ah yep we would have to update the docs for all the packages in the PR description (i realized i missed klaviyo and added the upstream affected packages). we should also update the shopify transform package

File: macros/union_data.sql
Comment (fivetran-joemarkiewicz) [inline macros/union_data.sql]: Could this maybe be misleading if customers believe this to be a native dbt variable? Should we consider renaming this to be `fivetran__remove_empty_table_warnings` or something like that?

File: macros/union_data.sql
Comment (fivetran-joemarkiewicz) [inline macros/union_data.sql]: What are your thoughts on adding the package information to the warning? I tested this locally where I was doing this union behavior for three packages and found it to be quite confusing which package was doing what with the staging models.

I found we can add the following to add a bit more details.
```sql
{{ exceptions.warn("\n\nPlease be aware: The " ~ table_identifier ~ " table was not found in your destination. The " ~ default_schema ~ " package will create a completely empty staging model as to not break downstream transformations. To turn off these warnings, set the `remove_empty_table_warnings` variable to False.\n") }}
```

Results:
```zsh
Please be aware: The discount_code table was not found in your destination. The shopify package will create a completely empty staging model as to not break downstream transformations. To turn off these warnings, set the `remove_empty_table_warnings` variable to False.

**AND**

Please be aware: The user_history table was not found in your destination. The instagram_business_pages package will create a completely empty staging model as to not break downstream transformations. To turn off these warnings, set the `remove_empty_table_warnings` variable to False.
```

It isn't fool proof as the naming won't match exactly what we have named the package (ie. we don't call the instagram_business package instagram_business_pages), but it may be enough to add more context.

I am trying to future proof this and I foresee us using this across a lot more packages, and feel this could get overwhelming for customers if they are unaware where these tables are being used. Thoughts?

**Review by fivetran-joemarkiewicz — COMMENTED** [review]
@fivetran-jamie this update is incredible and I think we should start to leverage this more widespread across our packages!

I have a few minor comments that I feel could possibly enhance the user experience based on my integration of a number of union macros in packages. Let me know your thoughts!

File: macros/union_data.sql
Comment (fivetran-jamie) [reply macros/union_data.sql]: yeah i was thinking of a fivetran prefix as well! i'll update

Comment (fivetran-jamie) [reply macros/union_data.sql]: ah now i remember -- i opted for no prefix at first to be consistent with the `has_defined_sources` variable we also reference in this macro (IMO, _that_ name is confusing and should have a `fivetran__` prefix as well, but i don't wanna break anything here 😬)

i still think adding the prefix to the new variable is the best move forward tho

File: macros/union_data.sql
Comment (fivetran-jamie) [reply macros/union_data.sql]: lmk whatcha think about what i added!

Comment (fivetran-joemarkiewicz) [reply macros/union_data.sql]: Looks great, I like the added details to the warning message!

File: macros/union_data.sql
Comment (fivetran-joemarkiewicz) [reply macros/union_data.sql]: Yeah let's flag this as something to update in the next breaking version of fivetran_utils. I like the update!

**Review by fivetran-joemarkiewicz — APPROVED** [review]
@fivetran-jamie great work on this PR! I just did my final review and have tested the following packages to ensure they work seamlessly on their current build of fivetran_utils and on the new one that will results from this update:
- Shopify
- Stripe
- Xero
- Instagram Business
- Klaviyo

I can confirm that they work without error and the updates highlighted here are functioning as intended. This looks great to move forward into the release branch!

**Pull request merged**
